packages/protocol/specs/reserve.spec

```diff
@@ -0,0 +1,27 @@
+pragma specify 0.1
+
+methods {
+  isSpender(address) returns bool envfree
+  isExchangeSpender(address) returns bool envfree
+  getExchangeAddress() returns address envfree
+  contractBalance() returns uint256 envfree
+}
+
+rule nonspenderCantReduceBalance(address a, method f) {
+  require(!isSpender(a));
+  require(!isExchangeSpender(a));
+  require(getExchangeAddress() != a);
+  uint256 balanceBefore = invoke contractBalance();
```

Comment:
you could omit the `invoke` here - if it reverts, value of `balanceBefore` is undefined.